### PR TITLE
DBZ-4627 Add backend service to get connector config

### PR DESCRIPTION
**Pull request by mdrillin**
Addition of the backend service to support fetching the connector config after the connector has been created.

### CODE REVIEW

**Comment by rk3rn3r** [comment]
@mdrillin like discussed on chat, could you update the PR to use the more specific Kafka Connect endpoint `/connectors/{connector-name}/config`?

**Comment by mdrillin** [comment]
@rk3rn3r - sorry, I'm not following what you mean - I thought what I have implemented here satisfies that

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java [hunk 1]
```diff
@@ -447,4 +447,50 @@ public Response deleteConnector(
...
+    public Response getConnectorConfig(
+            @PathParam("cluster") int cluster,
+            @PathParam("connector-name") String connectorName
+    ) throws KafkaConnectClientException, KafkaConnectException {
+        URI kafkaConnectURI = KafkaConnectClientFactory.getKafkaConnectURIforCluster(cluster);
+        KafkaConnectClient kafkaConnectClient = KafkaConnectClientFactory.getClient(cluster);
+
+        Map<String, String> connectorConfig = Collections.emptyMap();
+        try {
+            var connectorInfo = kafkaConnectClient.getConnectorInfo(connectorName);
```
[the remark this answers is not in the document]
Comment (mdrillin) [reply backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java]: Thanks @rk3rn3r - I updated the PR

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorURIs.java [hunk 2]
```diff
@@ -20,6 +20,7 @@
     public static final String CONNECTOR_RESUME_ENDPOINT = "/connector/{cluster}/{connectorname}/resume";
     public static final String CONNECTOR_RESTART_ENDPOINT = "/connector/{cluster}/{connectorname}/restart";
     public static final String CONNECTOR_TASK_RESTART_ENDPOINT = "/connector/{cluster}/{connectorname}/task/{tasknumber}/restart";
+    public static final String CONNECTOR_CONFIG_ENDPOINT = "/connectors/{cluster}/{connector-name}/config";
```
Comment (rk3rn3r) [inline backend/src/main/java/io/debezium/configserver/rest/ConnectorURIs.java:23]: Could you maybe use the same naming schema like in the other URLs? 
```suggestion
    public static final String CONNECTOR_CONFIG_ENDPOINT = "/connectors/{cluster}/{connectorname}/config";
```

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java [hunk 3]
```diff
@@ -447,4 +447,49 @@ public Response deleteConnector(
...
+    @APIResponse(
+            responseCode = "503",
+            description = "Exception while trying to connect to the selected Kafka Connect cluster",
+            content = @Content(
+                    mediaType = MediaType.APPLICATION_JSON,
+                    schema = @Schema(implementation = ServerError.class)
+            ))
+    public Response getConnectorConfig(
+            @PathParam("cluster") int cluster,
+            @PathParam("connector-name") String connectorName
```
Comment (rk3rn3r) [inline backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java:477]:
```suggestion
            @PathParam("connectorname") String connectorName
```

File: backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java [hunk 4]
```diff
@@ -46,6 +47,11 @@
     @Produces("application/json")
     ConnectConnectorStatusResponse getConnectorStatus(@PathParam("connector-name") String connectorName) throws ProcessingException, IOException;
 
+    @GET
+    @Path("/connectors/{connector-name}/config")
```
Comment (rk3rn3r) [inline backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java:51]:
```suggestion
    @Path("/connectors/{connectorname}/config")
```

File: backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java [hunk 5]
```diff
@@ -46,6 +47,11 @@
...
+    @Produces("application/json")
+    Map<String,String> getConnectorConfig(@PathParam("connector-name") String connectorName) throws ProcessingException, IOException;
```
Comment (rk3rn3r) [inline backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java:53]:
```suggestion
    Map<String,String> getConnectorConfig(@PathParam("connectorname") String connectorName) throws ProcessingException, IOException;
```

Comment (rk3rn3r) [inline backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java:53]: Missing comma (see CI error).

```suggestion
    Map<String,String> getConnectorConfig(@PathParam("connector-name"), String connectorName) throws ProcessingException, IOException;
```

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java [hunk 6]
```diff
@@ -447,4 +447,49 @@ public Response deleteConnector(
         }
         return deleteResponse;
     }
+
+    @Path(ConnectorURIs.CONNECTOR_CONFIG_ENDPOINT)
```
Comment (rk3rn3r) [inline backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java:451]: This method is just a proxy method for Kafka Connect and should therefore go to `KafkaConnectResource` class.

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java [hunk 7]
```diff
@@ -447,4 +447,49 @@ public Response deleteConnector(
...
+    public Response getConnectorConfig(
+            @PathParam("cluster") int cluster,
+            @PathParam("connectorname") String connectorName
+    ) throws KafkaConnectClientException, KafkaConnectException {
+        URI kafkaConnectURI = KafkaConnectClientFactory.getKafkaConnectURIforCluster(cluster);
+        KafkaConnectClient kafkaConnectClient = KafkaConnectClientFactory.getClient(cluster);
+
+        Map<String, String> connectorConfig = Collections.emptyMap();
+        try {
+            connectorConfig = kafkaConnectClient.getConnectorConfig(connectorName);
```
Comment (rk3rn3r) [inline backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java:484]: Could you please check the response headers of that response? I think you have to create a response copy with `Response.from()` like in other methods here to set a proper response content-type.

**Comment by rk3rn3r** [comment]
I added some more comments inline, but the basis looks already good.
Could you also add some integration tests to `ConnectorResourceIT` for that new endpoint (positive and negative tests).

File: backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java (hunk 5 above)
Comment (mdrillin) [reply backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java:53]: the error was referring to the lack of spacing in Map<String,String>.  Will fix that

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java [hunk 8]
```diff
@@ -447,4 +447,49 @@ public Response deleteConnector(
...
+                    schema = @Schema(implementation = ServerError.class)
+            ))
+    public Response getConnectorConfig(
+            @PathParam("cluster") int cluster,
+            @PathParam("connectorname") String connectorName
+    ) throws KafkaConnectClientException, KafkaConnectException {
+        URI kafkaConnectURI = KafkaConnectClientFactory.getKafkaConnectURIforCluster(cluster);
+        KafkaConnectClient kafkaConnectClient = KafkaConnectClientFactory.getClient(cluster);
+
+        Map<String, String> connectorConfig = Collections.emptyMap();
```
[the remark this answers is not in the document]
Comment (mdrillin) [reply backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java:482]: right - will fix

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java (hunk 7 above)
Comment (mdrillin) [reply backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java:484]: I think the doc [here](https://docs.confluent.io/platform/current/connect/references/restapi.html#get--connectors-(string-name)-config) says the response is a map?  If not, can you point me to an example to fix?

File: backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java (hunk 5 above)
Comment (rk3rn3r) [reply backend/src/main/java/io/debezium/configserver/rest/client/KafkaConnectClient.java]: Ah, yes. Sorry!

File: backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java (hunk 7 above)
Comment (rk3rn3r) [reply backend/src/main/java/io/debezium/configserver/rest/ConnectorResource.java:484]: Just search for `Response.from` in this class. We have some other proxy methods that forward the original response and re-append the proper content-type.

File: backend/src/test/java/io/debezium/configserver/LifecycleResourceIT.java [hunk 9]
```diff
@@ -176,4 +177,36 @@ public void testRestartPostgresConnectorTaskWithInvalidTaskNumber() {
                 .body("message", is("Unknown task: " + connectorName + "-" + invalidTaskNumber));
     }
 
+    @Test
+    public void testGetPostgresConnectorConfigNotFound() {
```
Comment (rk3rn3r) [inline backend/src/test/java/io/debezium/configserver/LifecycleResourceIT.java:181]: Tests for `ConnectorResource` methods are supposed to be in `ConnectorResourceIT`.
Lifecycle is more related to start, stop, delete of tasks. ConnectorResource is the right place, but could you move the test to the matching ConnectorResourceIT?

Comment (mdrillin) [reply backend/src/test/java/io/debezium/configserver/LifecycleResourceIT.java:181]: yeah after thinking about it, realized it did not really fit the purpose of the other LifecycleResourceITs.  I will move them

Comment (mdrillin) [reply backend/src/test/java/io/debezium/configserver/LifecycleResourceIT.java]: The PR has been updated
